design/tla/Task.tla

```diff
@@ -0,0 +1,760 @@
+TransitionTableOK ==
+  \A << s1, s2 >> \in Transition :
+    \/ s1 = null
+    \/ s2 = null
+    \/ s1 \preceq s2
+
+ASSUME TransitionTableOK  \* Note: ASSUME means ``check'' to TLC
+
+(* An action in which all transitions were valid. *)
+TransitionOK ==
+  \A id \in TaskId :
+     LET s1 == IF id \in DOMAIN tasks  THEN State(id)  ELSE null
+         s2 == IF id \in DOMAIN tasks' THEN State(id)' ELSE null
+     IN
+     << s1, s2 >> \in Transition
+
+(* The orchestrator never changes the state of a task (it only creates new ones, or sets `desired_state'). *)
+OrchestratorPreservesState ==
+  Orchestrator => \A i \in DOMAIN tasks \intersect DOMAIN tasks' : State(i) = State(i)'
+
+(* Some of the expressions below are ``temporal formulas''. Unlike state expressions and actions,
+   these look at a complete behaviour (sequence of states). Summary of notation:
+
+   [] means ``always''. e.g. []x=3 means that `x = 3' in all states.
+
+   <> means ``eventually''. e.g. <>x=3 means that `x = 3' in some state.
+
+   `x=3' on its own means that `x=3' in the initial state.
+*)
+
+\* A temporal formula that checks transitions
+TransitionsOK ==
+  /\ [][TransitionOK]_vars      \* Every step satisfies TransitionOK (or leaves `vars' unchanged)
+  /\ [][OrchestratorPreservesState]_vars
+
+(* Every service has the right number of running tasks (the system is in the desired state). *)
+InDesiredState ==
+  \A sid \in DOMAIN services :
+    LET runningTasks  == { i \in TasksOf(sid) : State(i) = running }
+        nRunning      == Cardinality(runningTasks)
+    IN
+    IF IsReplicated(sid) THEN
+       /\ nRunning = services[sid].replicas
+    ELSE
+       /\ IsGlobal(sid)
+       \* We have as many tasks as nodes:
+       /\ nRunning = Cardinality(Node)
+       \* We have a task for every node:
+       /\ { tasks[i].node : i \in runningTasks } = Node
+
+(* The main property we want to check.
+
+   []<> means ``always eventually'' (``infinitely-often'')
+
+   <>[] means ``eventually always'' (always true after some point)
+
+   This temporal formula says that if we only experience a finite number of
+   problems then the system will eventually settle on InDesiredState.
+*)
+EventuallyAsDesired ==
+  \/ []<> <<User>>_vars              \* Either the user keeps changing the configuration,
+  \/ []<> <<RestartTask>>_vars       \* or restarting/updating tasks,
+  \/ []<> <<WorkerDown>>_vars        \* or workers keep failing,
+  \/ []<> <<RejectTask>>_vars        \* or workers keep rejecting tasks,
+  \/ []<> <<ContainerExit>>_vars     \* or the containers keep exiting,
+  \/ <>[] InDesiredState             \* or we eventually get to the desired state and stay there.
+
+=============================================================================
+\* Modification History
+\* Last modified Thu Apr 19 14:55:46 BST 2018 by tal
+\* Created Mon Apr 16 11:31:19 BST 2018 by tal
```

Comment:
this is a dumb nit, but wouldn't the modification history just be tracked in git? Or does this history get baked into the PDF that a TLA+ file creates?